ramips: add support for TP-Link EX220 v1

dcampos opened the pull request:
This device is very similar, if not identical, to the TP-Link AX23 v1 but is targeted at service providers and features a completely different flash layout.

Hardware
--------

CPU:    MediaTek MT7621 DAT
RAM:    128MB DDR3 (integrated)
FLASH:  16MB SPI-NOR
WiFi:   MediaTek MT7905 + MT7975 (2.4 / 5 DBDC) 802.11ax
SERIAL: 115200 8N1
        LEDs - (3V3 - GND - RX - TX) - ETH ports

Installation
------------

Flashing is only possible via a serial connection using the sysupgrade image; the factory image must be signed. You can flash the sysupgrade image directly through the U-Boot console, or preferably, by booting the initramfs image and flashing with the sysupgrade command. Follow these steps for sysupgrade flashing:

1. Establish a UART serial connection.
2. Set up a TFTP server at 192.168.0.2 and copy the initramfs image there.
3. Power on the device and press any key to interrupt normal boot.
4. Load the initramfs image using tftpboot.
5. Boot with bootm.
6. If you haven't done so already, back up all stock mtd partitions.
7. Copy the sysupgrade image to the router.
8. Flash OpenWrt through either LuCI or the sysupgrade command. Remember not to attempt saving settings.

Revert to stock firmware
------------------------

Flash stock firmware via OEM web-recovery mode. If you don't have access to the stock firmware image, you will need to restore the firmware partition backed up earlier.

Web-Recovery
------------

The router supports an HTTP recovery mode:

1. Turn off the router.
2. Press the reset button and power on the device.
3. When all LEDs start flashing, release reset and quickly press it again.

The interface is reachable at 192.168.0.1 and supports installation of the OEM factory image. Note that flashing OpenWrt this way is not possible, as mentioned above.

raenye commented on target/linux/ramips/dts/mt7621_tplink_ex220-v1.dts:
```diff
@@ -0,0 +1,233 @@
+};
+
+&rom_file {
+	compatible = "nvmem-cells";
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	macaddr_rom_file_f100: macaddr@f100 {
+		reg = <0xf100 0x6>;
+	};
```
```suggestion
	nvmem-layout {
		compatible = "fixed-layout";
		#address-cells = <1>;
		#size-cells = <1>;

		macaddr_rom_file_f100: macaddr@f100 {
			compatible = "mac-base";
			reg = <0xf100 0x6>;
			#nvmem-cell-cells = <1>;
		};
	};
```

Also, all of this can move above, within `partitions`.

raenye commented on target/linux/ramips/dts/mt7621_tplink_ex220-v1.dts:
```diff
@@ -0,0 +1,233 @@
+};
+
+&gmac0 {
+	nvmem-cells = <&macaddr_rom_file_f100>;
```
```suggestion
	nvmem-cells = <&macaddr_rom_file_f100 0>;
```

raenye commented on target/linux/ramips/dts/mt7621_tplink_ex220-v1.dts:
```diff
@@ -0,0 +1,233 @@
+	label = "wan";
+	phy-handle = <&ethphy4>;
+
+	nvmem-cells = <&macaddr_rom_file_f100>;
+	nvmem-cell-names = "mac-address";
+	mac-address-increment = <1>;
```
```suggestion
	nvmem-cells = <&macaddr_rom_file_f100 1>;
	nvmem-cell-names = "mac-address";
```

raenye commented on target/linux/ramips/dts/mt7621_tplink_ex220-v1.dts:
```diff
@@ -0,0 +1,233 @@
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&radio 0x0>;
+		nvmem-cells = <&macaddr_rom_file_f100>;
```
```suggestion
		nvmem-cells = <&macaddr_rom_file_f100 0>;
```

raenye commented on target/linux/ramips/dts/mt7621_tplink_ex220-v1.dts:
```diff
@@ -0,0 +1,233 @@
+#include "mt7621.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
```
```suggestion
#include <dt-bindings/input/input.h>
#include <dt-bindings/leds/common.h>
```

This is required for `LED_COLOR_ID_XXX` and `LED_FUNCTION_XXX` below.

dcampos commented:
@raenye Thanks for the feedback! I've adjusted the pull request accordingly, and everything seems to be working just fine.

raenye commented:
BTW `10_fix_wifi_mac` shouldn't be needed now (and you kept LED node names instead of having `led-XXX`).

dcampos commented:
> BTW `10_fix_wifi_mac` shouldn't be needed now

Thanks, will try that.

> (and you kept LED node names instead of having `led-XXX`).

Could you elaborate on that? Do you mean I should change wifi2g, wifi5g to `led-wif2g`, `led-wif5g`, and so on?

dcampos commented:
I've applied the changes concerning the LED node names. 

Regarding `10_fix_wifi_mac`, I tried to remove it, but it caused both 2.4 Ghz and 5 Ghz wi-fi to be assigned the same MAC address. I'm not sure if there is any additional adjustments needed in order to be able to remove the EX220 section from `10_fix_wifi_mac`. Any guidance on this would be greatly appreciated.

BTW, stock firmware uses LAN MAC for 2.4 Ghz and LAN+2 for 5 Ghz.

raenye commented:
> Any guidance on this would be greatly appreciated.

The problem is that a [patch](https://patchwork.kernel.org/project/linux-wireless/list/?series=771380) is needed to make it possible to specific two different MAC addresses for a single DBDC WiFi IC. Once it is accepted, it'll be a quick DTS fix, but for now I guess the best way forward is to set the 2.4 GHz MAC in the DTS and the 5 GHz MAC in `10_fix_wifi_mac`.

Perhaps leave a comment in the dts saying 5 GHz MAC is set in `10_fix_wifi_mac` since this is a DBDC chip.

csharper2005 commented on target/linux/ramips/dts/mt7621_tplink_ex220-v1.dts:
```diff
@@ -0,0 +1,245 @@
+		led-wps {
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_WPS;
+			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
```
Would you like to sort the leds and the buttons by their gpio numbers?

dcampos replied:
Done.

csharper2005 commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -2369,6 +2369,19 @@ define Device/tplink_er605-v2
+  DEVICE_PACKAGES := kmod-mt7915-firmware -uboot-envtools
+  TPLINK_BOARD_ID := EX220-V1
+  KERNEL_LOADADDR := 0x82000000
+  KERNEL := kernel-bin | relocate-kernel 0x80001000 | lzma | fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb
```
Would be nice to wrap this long (>75 chars) line.

dcampos replied:
Thanks for pointing it out. Fixed.

dcampos committed:
ramips: add support for TP-Link EX220 v1

This device is very similar, if not identical, to the TP-Link AX23 v1
but is targeted at service providers and features a completely different
flash layout.

Hardware
--------

CPU:    MediaTek MT7621 DAT
RAM:    128MB DDR3 (integrated)
FLASH:  16MB SPI-NOR
WiFi:   MediaTek MT7905 + MT7975 (2.4 / 5 DBDC) 802.11ax
SERIAL: 115200 8N1
        LEDs - (3V3 - GND - RX - TX) - ETH ports

Installation
------------

Flashing is only possible via a serial connection using the sysupgrade
image; the factory image must be signed. You can flash the sysupgrade
image directly through the U-Boot console, or preferably, by booting the
initramfs image and flashing with the sysupgrade command. Follow these
steps for sysupgrade flashing:

1. Establish a UART serial connection.
2. Set up a TFTP server at 192.168.0.2 and copy the initramfs image
   there.
3. Power on the device and press any key to interrupt normal boot.
4. Load the initramfs image using tftpboot.
5. Boot with bootm.
6. If you haven't done so already, back up all stock mtd partitions.
7. Copy the sysupgrade image to the router.
8. Flash OpenWrt through either LuCI or the sysupgrade command. Remember
   not to attempt saving settings.

Revert to stock firmware
------------------------

Flash stock firmware via OEM web-recovery mode. If you don't have access
to the stock firmware image, you will need to restore the firmware
partition backed up earlier.

Web-Recovery
------------

The router supports an HTTP recovery mode:

1. Turn off the router.
2. Press the reset button and power on the device.
3. When all LEDs start flashing, release reset and quickly press it
   again.

The interface is reachable at 192.168.0.1 and supports installation of
the OEM factory image. Note that flashing OpenWrt this way is not
possible, as mentioned above.

Signed-off-by: Darlan Pedro de Campos <darlanpedro@gmail.com>
